MidCamp/midcamp
# Eventbrite integration

# Description

## Eventbrite Events module

https://github.com/Brian-Clement/eventbrite_events

Created a basic module for integrating Eventbrite.  From a high level, the module:

- Stores Eventbrite OAuth token in a config form
- Connects with Eventbrite via `GuzzleHttp` service
- Creates a content entity for "Eventbrite Attendee" to store attendee information.
- Creates a bundle for "Eventbrite Event" nodes which drive most of the functionality
    - Upon node presave, if an Eventbrite Event ID has been provided to the node, a call to the API is made to check and save the event status to a node field
    - Upon node save, if an Eventbrite Event ID has been provided, a call to the API is made to fetch the event attendees, which are stored as Eventbrite Attendee nodes
- Periodically (on cron) checks for new attendees for any active event nodes in the Drupal site

## Updates to MidCamp

- Enables Eventbrite Events module
- Adds Eventbrite OAuth token to config ignore to keep it out of the repo
- Adds "Attendees" view to display pulling 2019 attendees by tickets `MidCamp 2019 Regular Admission (March 21-22) ` and `MidCamp 2019 Individual Sponsorship (March 21-22)`
    - _NOTE: Requires an Eventbrite Event node with the MidCamp 2019 event ID_
- Adds Individual Sponsors block to the 2019 Sponsors page (also creates a standalone page at `/2019/individual-sponsors`)
- Adds template suggestion for `views_view_fields`
- Adds custom template for `views_view_fields` to reflect existing patterns for displaying lists of people

# To Test

Visit https://nginx-midcamp-org-feature-eventbrite-integration.us.amazee.io/2019/attendees to see the list of current MidCamp attendees.
Visit https://varnish-midcamp-org-feature-eventbrite-integration.us.amazee.io/2019/sponsors to see the individual sponsors listed at the bottom of the Sponsor page

## Locally

- Install dependencies with `composer install`
- Import configuration with `drush cim -y`
- Add the Eventbrite OAuth token at http://midcamp.org.docker.amazee.io/admin/config/eventbrite_events/settings (get this from the Eventbrite account or see me)
- Create an Eventbrite Event node at http://midcamp.org.docker.amazee.io/node/add/eventbrite_event and use the event ID for MidCamp 2019
- Upon saving, visit http://midcamp.org.docker.amazee.io/admin/structure/eventbrite_events_attendee and observe attendees have been synced

# Screenshots

![image](https://user-images.githubusercontent.com/4048700/51802623-8b65be00-2211-11e9-9048-e107d5dd8ce6.png)
Attendee list including ticket type and status.

![image](https://user-images.githubusercontent.com/4048700/51843707-31c8c680-22d9-11e9-8f32-1aaa2d4ce1d3.png)
Attendee display.

# To Do

## Eventbrite Events module

- [x] Check/update status on cron
- [ ] Disable form fields populated by Eventbrite
- [x] Tag a release to pin to

## Locally

- [x] Pin to a release of the eventbrite_events module
- [x] Menu link moved to "About" (consistent with 2016 architecture) 
- [x] Placeholder image for attendees without an image
- [ ] Verbiage for Attendees page
- [x] Individual Sponsor page (currently none)

# Questions

- Do we want a specific/different appearance for the attendee list?  Currently it is modeled largely after the Organizers page from last year (https://www.midcamp.org/2018/organizers). 
- Do we want default images to appear for users who do not have one?  See https://2016.midcamp.org/attendees.html for the last implementation like this.  For now I am using the Hatter logo.
- An individual sponsor page can be built easily using this approach, but there currently are none.  Do we want that included here knowing it will be empty for now?

# Notes

Looking for feedback on the approach and ways to improve it.  Please feel free to submit pull requests/issues to https://github.com/Brian-Clement/eventbrite_events.
